### PR TITLE
Fix panic during shoot spec and status check.

### LINKED ISSUE #27: invalid memory address or nil pointer dereference"
**Describe the bug:**

```
$ k -n garden get po dependency-watchdog-probe-5699594f8b-z8cd4
NAME                                         READY   STATUS    RESTARTS   AGE
dependency-watchdog-probe-5699594f8b-z8cd4   1/1     Running   4          101m
```

```
$ k -n garden logs dependency-watchdog-probe-5699594f8b-z8cd4 -p

I0924 09:18:44.628382       1 scaler.go:72] Update event on cluster: shoot--it--tmghj-0da
E0924 09:18:44.715691       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 62 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x18bbae0, 0x2b49eb0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x18bbae0, 0x2b49eb0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/gardener/dependency-watchdog/pkg/scaler.shootHibernationStateChanged(0xc000972bc0, 0xc0000b8d00, 0xa)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/scaler/utils.go:88 +0x341
github.com/gardener/dependency-watchdog/pkg/scaler.NewController.func1(0x1aa80e0, 0xc000972bc0, 0x1aa80e0, 0xc0000b8d00)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/scaler/scaler.go:79 +0xff
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/controller.go:205
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x0, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:656 +0x188
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc00dcc9dd8, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:292 +0x51
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:652 +0x79
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0003e6740)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00dcc9f40, 0xdf8475800, 0x0, 0x42da01, 0xc0005c6cc0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000367a00)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:650 +0x9b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0004c18e0, 0xc00035d290)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x59
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x17046a1]

goroutine 62 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x18bbae0, 0x2b49eb0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/gardener/dependency-watchdog/pkg/scaler.shootHibernationStateChanged(0xc000972bc0, 0xc0000b8d00, 0xa)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/scaler/utils.go:88 +0x341
github.com/gardener/dependency-watchdog/pkg/scaler.NewController.func1(0x1aa80e0, 0xc000972bc0, 0x1aa80e0, 0xc0000b8d00)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/scaler/scaler.go:79 +0xff
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/controller.go:205
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x0, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:656 +0x188
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc00dcc9dd8, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:292 +0x51
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:652 +0x79
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0003e6740)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00dcc9f40, 0xdf8475800, 0x0, 0x42da01, 0xc0005c6cc0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000367a00)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:650 +0x9b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0004c18e0, 0xc00035d290)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x59
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
```

**Anything else we need to know?**:

dependency-watchdog version: `v0.6.0`

### PR DESCRIPTION
**Pull request by amshuman-kr**
**What this PR does / why we need it**:
Fix panic during shoot spec and status check.

**Which issue(s) this PR fixes**:
Fixes #27 

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Fix panic during shoot spec and status check.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Thanks for the PR!

**Comment by amshuman-kr** [comment]
@rfranzke Should I cherry-pick this for a hotfix release of a main `v0.7.0` will be fine?

File: pkg/scaler/utils.go [hunk 1]
```diff
@@ -79,15 +80,23 @@ func shootHibernationStateChanged(old, new *gardenerv1alpha1.Cluster) bool {
 		return false
 	}
 
+	return doCheckShootHibernationStateChanged(oldShoot, newShoot)
+}
+
+func doCheckShootHibernationStateChanged(oldShoot, newShoot *gardencorev1beta1.Shoot) bool {
```
Comment (rfranzke) [inline pkg/scaler/utils.go:86]: What about

```golang
func doCheckShootHibernationStateChanged(oldShoot, newShoot *gardencorev1beta1.Shoot) bool {
  return oldShoot.Status.IsHibernated != newShoot.Status.IsHibernated || gardencorev1beta1helper.HibernationIsEnabled(oldShoot) != gardencorev1beta1helper.HibernationIsEnabled(newShoot)
}
```

https://github.com/gardener/gardener/blob/79ed94b076c662bc8b49b81fd076b704f7c37d8c/pkg/apis/core/v1beta1/helper/helper.go#L585-L587

**Comment by rfranzke** [comment]
@amshuman-kr I don't care too much, TBH :) Technically, it's rather a hotfix, so cherry-picking makes sense, but you are the component owner, so it's your decision :)

File: pkg/scaler/utils_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	yes = true
+	no  = false
+)
+
+var _ = FDescribeTable("doCheckShootHibernationStateChanged", func(oldSpec *gardencorev1beta1.Hibernation, oldStatus bool, newSpec *gardencorev1beta1.Hibernation, newStatus, expectChanged bool) {
```
[the remark this answers is not in the document]
Comment (amshuman-kr) [reply pkg/scaler/utils_test.go:28]: Oops. Thanks for catching it :-)

File: pkg/scaler/utils.go (hunk 1 above)
Comment (ialidzhikov) [reply pkg/scaler/utils.go:86]: +1

**Comment by amshuman-kr** [comment]
> I don't care too much, TBH :) Technically, it's rather a hotfix, so cherry-picking makes sense, but you are the component owner, so it's your decision :)

I will the _right thing_ and create a hotfix :-)

@rfranzke @ialidzhikov I have made the suggested changed and rebased. Can you PTAL?

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Comment by amshuman-kr** [comment]
@ggaurav10 Could you please have a look? I need your review to be able to merge.

**Review by ggaurav10 — APPROVED** [review]
lgtm

**Pull request merged**